testing/matrix-sdk-test/src/event_builder.rs

```diff
@@ -0,0 +1,207 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Unit tests (based on private methods) for the timeline API.
+
+use std::{
+    collections::BTreeMap,
+    sync::atomic::{AtomicU64, Ordering::SeqCst},
+};
+
+use ruma::{
+    events::{
+        receipt::{Receipt, ReceiptEventContent, ReceiptThread, ReceiptType},
+        relation::Annotation,
+        AnySyncTimelineEvent, AnyTimelineEvent, MessageLikeEventContent,
+        RedactedMessageLikeEventContent, RedactedStateEventContent, StateEventContent,
+    },
+    serde::Raw,
+    server_name, EventId, MilliSecondsSinceUnixEpoch, OwnedEventId, OwnedUserId, RoomId, UserId,
+};
+use serde_json::{json, Value as JsonValue};
+
+#[derive(Default)]
+pub struct EventBuilder {
+    next_ts: AtomicU64,
+}
+
+impl EventBuilder {
+    pub fn new() -> EventBuilder {
+        Self { next_ts: AtomicU64::new(0) }
+    }
+
+    pub fn next_server_ts(&self) -> MilliSecondsSinceUnixEpoch {
+        MilliSecondsSinceUnixEpoch(
+            self.next_ts
+                .fetch_add(1, SeqCst)
+                .try_into()
+                .expect("server timestamp should fit in js_int::UInt"),
+        )
+    }
+
+    /// Set the next server timestamp.
+    ///
+    /// Timestamps will continue to increase by 1 (millisecond) from that value.
+    pub fn set_next_ts(&self, value: u64) {
+        self.next_ts.store(value, SeqCst);
+    }
+
+    pub fn make_message_event_with_id<C: MessageLikeEventContent>(
+        &self,
+        sender: &UserId,
+        room_id: &RoomId,
+        event_id: &EventId,
+        content: C,
+    ) -> Raw<AnyTimelineEvent> {
+        timeline_event!({
+            "type": content.event_type(),
+            "content": content,
+            "event_id": event_id,
+            "sender": sender,
+            "room_id": room_id,
+            "origin_server_ts": self.next_server_ts(),
+        })
+    }
+
+    pub fn make_sync_message_event<C: MessageLikeEventContent>(
+        &self,
+        sender: &UserId,
+        content: C,
+    ) -> Raw<AnySyncTimelineEvent> {
+        let event_id = EventId::new(server_name!("dummy.server"));
+        self.make_sync_message_event_with_id(sender, &event_id, content)
+    }
+
+    pub fn make_sync_message_event_with_id<C: MessageLikeEventContent>(
+        &self,
+        sender: &UserId,
+        event_id: &EventId,
+        content: C,
+    ) -> Raw<AnySyncTimelineEvent> {
+        sync_timeline_event!({
+            "type": content.event_type(),
+            "content": content,
+            "event_id": event_id,
+            "sender": sender,
+            "origin_server_ts": self.next_server_ts(),
+        })
+    }
```

Comment:
The existence and similarity of those two methods suggest a builder pattern, if you're so inclined (we could also revisit later when needs be).

Reply:
Ah but that requires separate finalization, which makes test code longer again :(